From 1d2da035115fbc1ea6d539acf32180f504ef676a Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:14:38 -0400
Subject: [net] icmp6: Call skb_checksum_validate

Message-id: <1409652920-431-16-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90141
O-Subject: [PATCH 15/57] [RHEL7.1 net] icmp6: Call skb_checksum_validate
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

small merge conflict due to linebreak change in the
removed switch/case.

commit 39471ac8dde690bf944248e06bec32a4568cdd45
Author: Tom Herbert <therbert@google.com>
Date:   Wed May 7 16:52:29 2014 -0700

    icmp6: Call skb_checksum_validate

    Use skb_checksum_validate to verify checksum.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/icmp.c b/net/ipv6/icmp.c
index 11a2a74..2864cd3 100644
--- a/net/ipv6/icmp.c
+++ b/net/ipv6/icmp.c
@@ -686,21 +686,11 @@ static int icmpv6_rcv(struct sk_buff *skb)
  saddr = &ipv6_hdr(skb)->saddr;
  daddr = &ipv6_hdr(skb)->daddr;
 
- /* Perform checksum. */
- switch (skb->ip_summed) {
- case CHECKSUM_COMPLETE:
-  if (!csum_ipv6_magic(saddr, daddr, skb->len, IPPROTO_ICMPV6,
-         skb->csum))
-   break;
-  /* fall through */
- case CHECKSUM_NONE:
-  skb->csum = ~csum_unfold(csum_ipv6_magic(saddr, daddr, skb->len,
-          IPPROTO_ICMPV6, 0));
-  if (__skb_checksum_complete(skb)) {
-   LIMIT_NETDEBUG(KERN_DEBUG "ICMPv6 checksum failed [%pI6 > %pI6]\n",
-           saddr, daddr);
-   goto csum_error;
-  }
+ if (skb_checksum_validate(skb, IPPROTO_ICMPV6, ip6_compute_pseudo)) {
+  LIMIT_NETDEBUG(KERN_DEBUG
+          "ICMPv6 checksum failed [%pI6c > %pI6c]\n",
+          saddr, daddr);
+  goto csum_error;
  }
 
  if (!pskb_pull(skb, sizeof(*hdr)))
-- 
1.7.1